running-code/execute-package.md

```diff
@@ -0,0 +1,163 @@
+    jupytext_version: 1.16.4
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+(execute-package)=
+# How to Execute a Python Package
+
+In [How to Execute a Python Script](execute-script) you learned about two primary ways to execute a stand-alone Python script.
+There are two other ways to execute Python code as commands, both of which work for code that has been formatted as a package.
+
+1. You can execute modules using their import name
+2. A package can provide arbitraty command names that execute parts of themselves
+
+## 1. Executable modules
+
+In the [Pass your script to the Python command lesson](execute-script-pass-to-python) you learned that the `python` command can
+be passed a file path for execution. Alternatively you can also pass the name of a module, exactly as would be used after an `import`.
+In this case, Python will look up the module referenced in the
+[currently active environment](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/#create-and-use-virtual-environments),
+and will execute it as a script.
+
+This execution mode is performed with the `-m` flag, as in `python -m site`. It can be used in place of a file
+path, but cannot be used in combination with a path, as there can only be one executing module.
+
+:::{tip}
+These commands both do the same thing, but the latter is easier to remember and much more portable, as it doesn't rely
+on the local machine's file system details.
+
+```bash
+python ./.venv/lib/python3.12/site-packages/pip/__main__.py
+```
+
+```bash
+python -m pip
+```
+:::
+
+### Further exploration
+
+On your own or in small groups:
+
+* Install the `my_program.py` module from the [last lesson](execute-script-launch-command)
+* Try to get the same greeting as before using `python -m my_program`.
+
+**my_program.py** which prints out `✨ Hello from Python ✨` when executed as a script.
+
+```python
+#!/usr/bin/env python
+# The above line is a shebang and can take the place of typing python on the command line
+# This comment is below, because shebangs must be the first line!
+
+def shiny_hello():
+    print("\N{Sparkles} Hello from Python \N{Sparkles}")
+
+if __name__ == "__main__":
+    shiny_hello()
+```
+
+### Executable packages
+
+The `-m` flag as described above only works for Python modules (files), but does not work for Python (sub-)packages (directories).
+This means that you cannot execute a command using only the name of your package when it is structured to use directories
+
+Once your package grows, the top-level name `my_program` turns into a directory.
+(See [Python Package Structure](https://www.pyopensci.org/python-package-guide/package-structure-code/python-package-structure.html)
+for when and how to create a package structure).
+```
+project/
+└── src/
+    └── my_program/
+        ├── __init__.py
+        └── greeting.py
+```
+
+However, you can never execute a directory.
```

Comment:
you can though! if it has a `__main__.py` file (i think the "never" might be confusing).

Reply:
ya, probably don't need the never. I think I was coming at it like "the directory doesn't have any binary to load onto the CPU" but then say how a directory can be an execution target.

Reply:
also apparently if you run `python -m package.subpackage` and the subpackage only has an `__init__.py` file in it, then the `__init__.py` file runs, and maybe predictably when it has a `__main__.py` file it runs both. so we might also treat this section as like "for a package, you can think of the `__init__.py` file as the contents of a module and the `__main__.py` file as the `if __name__ == "__main__":` block. I think you already do this but i guess that is a more 1:1 comparison than i thought